20x speedup for BufferTextWriter

BrannonKing opened the pull request:
I posted this pull request once before, but I've now updated it for 2.2 and fixed the spacing. The JSON serializer uses the TextWriter's `Write(char[], int, int)` overload.

halter73 commented:
Previous PR: https://github.com/SignalR/SignalR/pull/3378

davidfowl commented on src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs:
```diff
@@ -200,10 +204,9 @@ internal void OpenConnection(IConnection connection, string data, CancellationTo
-                _request.Abort();
-            }
+            var request = _request;
+            if (request != null)
```
Keep the style

davidfowl commented on src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs:
```diff
@@ -107,7 +109,9 @@ internal void OpenConnection(IConnection connection, string data, CancellationTo
                 // This will no-op if the request is already finished.
-                ((IRequest)state).Abort();
+                var request = state as IRequest;
+                if (request != null)
```
When would it be null?

BrannonKing replied:
I don't know when it would be null. I was attempting to fix this issue: https://github.com/SignalR/SignalR/issues/3376 . I was also trying to make the file consistent; I believe that there are null checks for it elsewhere in the file.

davidfowl commented on src/Microsoft.AspNet.SignalR.Client/Transports/ServerSentEventsTransport.cs:
```diff
@@ -53,7 +53,9 @@ protected override void OnStartFailed()
 
-            _request.Abort();
+            var request = _request;
+            if (request != null)
```
When would it be null?

BrannonKing replied:
Issue https://github.com/SignalR/SignalR/issues/3376 shows that it can be null, but I don't know when.

moozzyk replied:
This should be a separate PR. The NRE happens if the disconnectToken is tripped before sending the connect request.

davidfowl commented on src/Microsoft.AspNet.SignalR.Core/Infrastructure/PerformanceCounterManager.cs:
```diff
@@ -438,6 +438,14 @@ public IPerformanceCounter LoadCounter(string categoryName, string counterName,
             // and when they are thrown. 
             try
             {
+                if (!PerformanceCounterCategory.Exists(categoryName))
```
We removed the Exists check for a reason. Please remove it.

BrannonKing replied:
I added the check as I have some developers on my team who disable the "just my code" option from time to time. This exception was causing them some grief. I don't believe that having the counters not installed is an unexpected scenario. Help me understand why the check is bad?

rjvdboon replied:
The Exists check can be very expensive, look at the Remarks section of the PerformanceCounterCategory.Exists member on msdn. It's much faster to try to open and fail than to check in advance.

davidfowl commented:
This definitely needs unit tests

BrannonKing commented:
I've added a unit test for verification of the method behavior, but not verified that the method is used when appropriate. I couldn't figure out a way to accomplish that latter goal. I also added braces to fix the styling.

moozzyk commented:
I believe this was ported to 2.2.1
